src/mixings/logit_sb_mixing.cc

```diff
@@ -0,0 +1,115 @@
+#include "logit_sb_mixing.h"
+
+#include <google/protobuf/stubs/casts.h>
+
+#include <Eigen/Dense>
+#include <stan/math/prim.hpp>
+
+#include "mixing_prior.pb.h"
+#include "mixing_state.pb.h"
+#include "src/utils/proto_utils.h"
+#include "src/utils/rng.h"
+
+void LogitSBMixing::initialize(const unsigned int n_clust /*= 1*/) {
+  if (prior == nullptr) {
+    throw std::invalid_argument("Mixing prior was not provided");
+  }
+  num_clusters = n_clust;
+  auto priorcast = cast_prior();
+  precision = stan::math::inverse_spd(
+    bayesmix::to_eigen(priorcast->normal_prior().var()));
+  initialize_state();
+}
+
+void LogitSBMixing::initialize_state() {
+  auto priorcast = cast_prior();
+  if (priorcast->has_normal_prior()) {
+    Eigen::VectorXd prior_vec =
+        bayesmix::to_eigen(priorcast->normal_prior().mean());
+    dim = prior_vec.size();
+    if (dim != precision.cols()) {
+      throw std::invalid_argument(
+        "Hyperparameters dimensions are not consisent");
+    }
+    if (priorcast->proposal_var() <= 0) {
+      throw std::invalid_argument("Proposal variance parameter must be > 0");
+    }
+    if (priorcast->step_size() <= 0) {
+      throw std::invalid_argument("Step size parameter must be > 0");
+    }
+    state.regression_coeffs = Eigen::MatrixXd(dim, num_clusters);
+    for (int i = 0; i < num_clusters; i++) {
+      state.regression_coeffs.col(i) = prior_vec;
+    }
+
+  } else {
+    throw std::invalid_argument("Unrecognized mixing prior");
+  }
+}
+
+void LogitSBMixing::update_state(
+    const std::vector<std::shared_ptr<AbstractHierarchy>> &unique_values,
+    const unsigned int n) {
+  // Langevin-Adjusted Metropolis-Hastings step
+  auto &rng = bayesmix::Rng::Instance().get();
+  Eigen::VectorXd state_c = state.regression_coeffs;
+  auto priorcast = cast_prior();
+  Eigen::VectorXd prior_mean = bayesmix::to_eigen(
+      priorcast->normal_prior().mean());
+  double prop_var = priorcast->proposal_var();
+  double step = priorcast->step_size();
+  // Rebuild allocations vector
+  std::vector<unsigned int> allocations;  // TODO
```

Comment:
The Durante-Rigon update mandates that we know whether all data points belong or not to a previous cluster than the one currently considered. Basically, we need the whole allocations vector. We can easily rebuild it from `unique_values` which is passed as argument, but this would take time which is basically wasted since we already have `BaseAlgorithm::allocations`. Should we also pass it as argument?

Reply:
I would go the easy way and just pass the allocations. We can store a pointer to the covariates during initialization, as you said those are fixed